pkg/util/winutil/process.go

```diff
@@ -0,0 +1,278 @@
+	}
+	return
+}
+
+// ReadProcessMemory wraps the Windows kernel.dll function of the same name
+// https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-readprocessmemory
+func ReadProcessMemory(h windows.Handle, from, to uintptr, count uint32) (bytesRead uint64, err error) {
+	var bytes uint64
+
+	r, _, e := procReadProcessMemory.Call(uintptr(h),
+		from,
+		to,
+		uintptr(count),
+		uintptr(unsafe.Pointer(&bytes)))
+
+	if r == 0 {
+		err = windows.GetLastError()
+		fmt.Printf("Error %v %v\n", err, e)
+		return
+	}
+	bytesRead = bytes
+	return
+}
+
+type peb32 struct {
+	Reserved1         [2]byte
+	BeingDebugged     byte
+	Reserved2         [1]byte
+	Reserved3         [2]uint32
+	Ldr               uint32
+	ProcessParameters uint32
+	// more fields...
+}
+
+type unicodeString32 struct {
+	length    uint16
+	maxLength uint16
+	buffer    uint32
+}
+type procParams32 struct {
+	Reserved1              [16]byte
+	Reserved2              [5]uint32
+	CurrentDirectoryPath   unicodeString32
+	CurrentDirectoryHandle uint32
+	DllPath                unicodeString32
+	ImagePath              unicodeString32
+	commandLine            unicodeString32
+	env                    uint32
+}
+
+func getCommandLineForProcess32(h windows.Handle) (cmdline string, err error) {
+	// get the pointer to the PEB
+	var procmem uintptr
+	size := unsafe.Sizeof(procmem)
+	err = NtQueryInformationProcess(h, ProcessWow64Information, uintptr(unsafe.Pointer(&procmem)), size)
+	if err != nil {
+		// this shouldn't happen because we already know we're asking about
+		// a 32 bit process.
+		return
+	}
+	var peb peb32
+	var read uint64
+	var toRead uint32
+	toRead = uint32(unsafe.Sizeof(peb))
+
+	read, err = ReadProcessMemory(h, procmem, uintptr(unsafe.Pointer(&peb)), toRead)
+	if err != nil {
+		return
+	}
+	if read != uint64(toRead) {
+		err = fmt.Errorf("Wrong amount of bytes read %v != %v", read, toRead)
+		return
+	}
+
+	// now go get the actual parameters
+	var pparams procParams32
+	pparamsSize := unsafe.Sizeof(pparams)
+
+	read, err = ReadProcessMemory(h, uintptr(peb.ProcessParameters), uintptr(unsafe.Pointer(&pparams)), uint32(pparamsSize))
+	if err != nil {
+		return
+	}
+	if read != uint64(pparamsSize) {
+		err = fmt.Errorf("Wrong amount of bytes read %v != %v", read, pparamsSize)
+		return
+	}
+	cmdlinebuffer := pparams.commandLine.buffer
+	cmdlinelen := pparams.commandLine.length
+	//cmdlinelen := 1024
+
+	finalbuf := make([]uint8, cmdlinelen+2)
+
+	read, err = ReadProcessMemory(h, uintptr(cmdlinebuffer), uintptr(unsafe.Pointer(&finalbuf[0])), uint32(cmdlinelen+2))
+	if err != nil {
+		return
+	}
+	if read != uint64(cmdlinelen+2) {
+		err = fmt.Errorf("Wrong amount of bytes read %v != %v", read, cmdlinelen+2)
+		return
+	}
+	cmdline = ConvertWindowsString(finalbuf)
+	return
+
+}
+
+// this definition taken from Winternl.h
+type unicodeString struct {
+	length    uint16
+	maxLength uint16
+	buffer    uintptr
+}
+
+type _rtlUserProcessParameters struct {
```

Comment:
Any reason for the different nomenclature style between the win32 and win64 variants? Consistency would be nice, unless we have a good reason (ie. this is shooting for windows API consistency and hence the names). I was kind of expecting `peb64` and `procParams64`.